compiler/rustc_middle/src/ty/consts/valtree.rs

```diff
@@ -16,9 +20,9 @@ use crate::ty::{self, Ty, TyCtxt};
 ///
 /// `ValTree` does not have this problem with representation, as it only contains integers or
 /// lists of (nested) `ValTree`.
-#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
 #[derive(HashStable, TyEncodable, TyDecodable)]
-pub enum ValTree<'tcx> {
+pub enum ValTreeKind<'tcx> {
```

Comment:
The `'tcx` lifetime is now not needed here anymore

Reply:
Its still used in `Branch` which now contains the new interned `ValTree` instead of the old/renamed `ValTreeKind`.

Reply:
Ah I see... is that part of what made everything more performant?

Reply:
Not really, all the relevant benchmarks use `ValTreeKind::Leaf` only. The main reason interning is faster is that the comparison `lhs_val.valtree() == rhs_val.valtree()` from the snippet linked in the OP is faster.

I just replaced the valtree in `Branch` with the interned one because it made the code simpler, but I suspect that it's also at least not slower.

Reply:
Kinda sad about the double indirection with `Box<[]>` and then the interned value, but we can golf that when we have actual use cases